### ISSUE TITLE
Move inventory information out of Kptfile

### ISSUE THREAD

**Issue by morgante** [kptdev/kpt#2399]
### Describe your problem

In kpt v1, inventory information was moved into the Kptfile. This makes the UX for developing/iterating on packages more painful.

As a package author, I frequently want to apply a package myself to test it locally. This requires that I `init` it and inventory information is pushed into the `Kptfile`. I do not want that inventory information committed to Git (as this package will be reused by others).

Inventory information should be moved out of the Kptfile.

**Comment by phanimarupaka** [comment]
@morgante I think the problem here is because package publisher might push the inventory information to Git. There are multiple ways of solving it. One way is to have a publish workflow in Kpt which will validate kpt packages before publishing.

Even if we move the inventory information out of Kptfile, the problem is not solved as the publishers might still push inventory file to Git.

**Comment by morgante** [comment]
@phanimarupaka Yes, the problem is that I might publish the inventory info to Git.

I don't want to learn a special kpt publish workflow. I just want to interact with Git like I usually do. Developers already have Git workflows and we shouldn't break that.

> Even if we move the inventory information out of Kptfile, the problem is not solved as the publishers might still push inventory file to Git.

Disagree. If we moved it into a `inventory` file it's trivial to `.gitignore` it.

**Comment by droot** [comment]
Actually core of the issue is that there are two types of information related to a package:

- Static metadata that includes resources, package information, files to ignore (part of krmignore etc.), subpackages and default customizations (part of pipeline if applicable)
- Runtime state which is typically associated with the instantiation of a package in a cluster

I think decoupling these type of info can potentially make it easier to support different workflows (publishing, even adopting an existing runtime state) but needs to be thought through.  publish command proposal had some other motivations as well (we documented that somewhere, need to dig that out). 

/cc @mortent

**Comment by morgante** [comment]
@droot Right, the core of the issue is a single file collapsing those two kinds of info.

> but needs to be thought through. publish command proposal had some other motivations as well (we documented that somewhere, need to dig that out).

If you want to *also* provide a publish command, that might be okay but I want to be clear I would not consider that a sufficient resolution to this issue. I should be able to use normal Git commands to interact with Kpt.

**Comment by mikebz** [comment]
What's the cost of doing this change?  This might also inform the decision.  I can see how decoupling those two things can be useful and we see some tools splitting transient information into a separate file.

**Comment by droot** [comment]
> Can you explain to me why it is necessary to have both pieces of information in a single file?

@mortent @seans3 might be more familiar with the reasoning here, but one possible guess is to avoid config sprawl (kptfile, krmignore, inventoryinfo,....)

We can definitely improve the package {dev and publishing} CUJ here. I have added this as one of the items to focus on.

**Comment by rustycl0ck** [comment]
I have the same problem.

I applied my package locally, made some updates to the `Kptfile`, applied again. After verifying that my package was working fine, I removed the `inventory` section for doing `git commit` and `git push`. And now I have no way to `destroy` the package which I forgot to uninstall earlier before committing my changes.

**Comment by mikebz** [comment]
@rquitales @yuwenma please close this once you are done with your current `kpt live apply` changes.

**Comment by PSanetra** [comment]
@rquitales @yuwenma Is this considered fixed since https://github.com/GoogleContainerTools/kpt/pull/2615 was merged?

**Comment by yuwenma** [comment]
> @rquitales @yuwenma Is this considered fixed since #2615 was merged?

It is not. http://github.com/GoogleContainerTools/kpt/pull/2802 has some major changes that need to go in as well. @rquitales Are you still working on this feature? If so, can you resolve @mortent comments and the code conflict?
